quickcache/quickcache.py

```diff
@@ -1,329 +1,67 @@
-
-          now as soon as request.couch_user._rev has changed,
-          the function will be recomputed
-
-    Features:
-        - In addition to caching in the default shared cache,
-          quickcache caches in memory for 10 seconds
-          (conceptually the length of a single request).
-          This can be overridden to a different number with memoize_timeout.
-
-        - In addition to varying on the arguments and the name of the function,
-          quickcache will also make sure to vary
-          on the _source code_ of your function.
-          That way if you change the behavior of the function, there won't be
-          any stale cache when you deploy.
-
-        - Can vary on any number of the function's parameters
-
-        - Does not by default vary on all function parameters.
-          This is because it is not in general obvious what it means
-          to vary on an object, for example.
-
-        - Allows you to vary on an attribute of an object,
-          multiple attrs of the same object, attrs of attrs of an object, etc
-
-        - Allows you to pass in a function as the vary_on arg which will get called
-          with the same args and kwargs as the function. It should return a list of simple
-          values to be used for generating the cache key.
-
-        Note on unicode and strings in vary_on:
-          When strings and unicode values are used as vary on parameters they will result in the
-          same cache key if and only if the string values are UTF-8 or ascii encoded.
-          e.g.
-          u'namé' and 'nam\xc3\xa9' (UTF-8 encoding) will result in the same cache key
-          BUT
-          u'namé' and 'nam\xe9' (latin-1 encoding) will NOT result in the same cache key
-
-
-    """
-    if cache and timeout:
-        raise ValueError('You can only use timeout '
-                         'when not overriding the cache')
-
-    timeout = timeout if timeout is not None else 5 * 60
-    memoize_timeout = memoize_timeout or 10
-
-    if cache is None:
-        caches = [
-            CacheWithTimeout(
-                django_caches['locmem'],
-                timeout=memoize_timeout)]
-        if timeout > 0:
-            caches.append(
-                CacheWithTimeout(
-                    django_caches['default'], timeout=timeout))
-        cache = TieredCache(caches)
-
-    def decorator(fn):
-        helper = helper_class(fn, vary_on=vary_on, cache=cache)
-
-        @functools.wraps(fn)
-        def inner(*args, **kwargs):
-            return helper(*args, **kwargs)
-
-        inner.clear = helper.clear
-        inner.get_cache_key = helper.get_cache_key
-        inner.prefix = helper.prefix
-        inner.get_cached_value = helper.get_cached_value
-
-        return inner
-
-    return decorator
+get_quickcache = QuickCache(
+    vary_on=Ellipsis,
+    cache=Ellipsis,
+    skip_arg=None,
+    helper_class=QuickCacheHelper,
+    assert_function=assert_function,
+).but_with
```

Comment:
This file is 100% new code; most of the "deleted" code in the diff was actually just moved to another file. `CacheWithTimeout` and `TieredCache` were moved verbatim to `cache_helpers.py` and `QuickCache` and `SkippableQuickCache` were merged into a single class and moved to `quickcache_helper.py`.

Where we previously exposed a `quickcache` decorator-maker function directly, we now expose `get_quickcache` which you can call to configure your decorator-maker. I arranged it so that these `QuickCache` objects, which are spruced-up `namedtuples`, can be used to generate a copy of themselves with modified settings by calling `but_with` (which is just an alias for `namedtuple`'s `._replace`). `get_quickcache` is then actually an alias for a preconfigured `QuickCache` instance's `but_with` method, i.e. "take this default object and replace values as you like".

It's a little roundabout, but was actually super helpful in making sure that the settings can be overridden at any point from your initial `quickcache` configuration to when you're using it as a decorator-make, and that the behavior is the same no matter where it's set. As you can see, the `__call__` method above (which is what's called when you do `@quickcache(...)`) first uses `.but_with` on the arguments you pass, which makes sure they're treated the same as if they were set on the object in the first place. (Like I said, it's a bit roundabout, but when I was toying around with other ways of doing this, it was by far the logically simplest/easiest to reason with while getting the desired behavior.)